File: factory-method/java/FactoryMethod.java
// 需要先定义产品
interface Product {
  void productOperation();
}
// 然后定义创建产品的生产者
interface Creator {
  Product factoryMethod();
  void creatorOperation();
}
// 具体产品的定义
class ConcreteProduct implements Product {

  @Override
  public void productOperation() {
    System.out.println("Product instance has been created");
  }
}
// 具体生产者的定义
class ConcreteCreator implements Creator {

  @Override
  public Product factoryMethod() {
    return new ConcreteProduct();
  }

  @Override
  public void creatorOperation() {
    System.out.println("Creator instance has been created");
  }
}

public class FactoryMethod {
  public static void main(String[] args) {
    Creator creator = new ConcreteCreator();
    Product product = creator.factoryMethod();
    creator.creatorOperation();
    product.productOperation();
  }
}

File: factory-method/python/FactoryMethod.py
#!/usr/bin/env python
# coding:utf8

from abc import abstractmethod, ABCMeta


class Product(metaclass=ABCMeta):
    @abstractmethod
    def productOperation(self):
        pass


class Creator(metaclass=ABCMeta):
    @abstractmethod
    def creatorOperaton(self):
        pass

    @abstractmethod
    def factoryMethod(self):
        pass


class ConcreteProduct(Product):
    def productOperation(self):
        print("Product instance has been created")


class ConcreteCreator(Creator):
    def creatorOperaton(self):
        print("Creator instance has been created")

    def factoryMethod(self):
        return ConcreteProduct()


if __name__ == "__main__":
    creator = ConcreteCreator()
    product = creator.factoryMethod()
    creator.creatorOperaton()
    product.productOperation()
